plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -0,0 +1,493 @@
+		"0x9E", "0x6A", "0x6C", "0x7D", "0x7E", "0x9D", "0x3C", "0x3F", "0x45", "0x46", "0x3D", "0x47", "0x4F", "0x56",
+		"0x66", "0x57", "0x85", "0xA5", "0xA6", "0x8F", "0x8C", "0x8D"}
+)
+
+// PowerStat plugin enables monitoring CPU utilization and power consumption on Intel platform.
+type PowerStat struct {
+	CPUMetrics []string        `toml:"cpu_metrics"`
+	Log        telegraf.Logger `toml:"-"`
+
+	fs   services.FileService
+	rapl services.RaplService
+	msr  services.MsrService
+
+	perCoreMetrics         bool
+	coreCPUFrequency       bool
+	c1CoreStateResidency   bool
+	c6CoreStateResidency   bool
+	processorBusyCycles    bool
+	coreCPUTemperature     bool
+	processorBusyFrequency bool
+	cpuInfo                map[string]*data.CPUInfo
+	skipFirstIteration     bool
+}
+
+// Init performs one time setup of the plugin.
+func (p *PowerStat) Init() error {
+	p.ParseCPUMetricsConfig()
+	err := p.verifyProcessor()
+	if err != nil {
+		return err
+	}
+	// Set perCoreMetrics flag only when there is at least one core metric enabled.
+	if p.c1CoreStateResidency || p.c6CoreStateResidency || p.processorBusyCycles || p.coreCPUTemperature ||
+		p.processorBusyFrequency || p.coreCPUFrequency {
+		p.perCoreMetrics = true
+		p.msr = services.NewMsrServiceWithFs(p.Log, p.fs)
+	}
+	p.rapl = services.NewRaplServiceWithFs(p.Log, p.fs)
+
+	return nil
+}
+
+// SampleConfig returns the default configuration of the plugin.
+func (p *PowerStat) SampleConfig() string {
+	return `
+  ## All global metrics are always collected by Intel PowerStat plugin.
+  ## User can choose which per-CPU metrics are monitored by the plugin in cpu_metrics array.
+  ## Empty array means no per-CPU specific metrics will be collected by the plugin - in this case only platform level
+  ## telemetry will be exposed by Intel PowerStat plugin.
+  ## Supported options:
+  ## "cpu_frequency", "cpu_c1_state_residency", "cpu_c6_state_residency", "cpu_busy_cycles", "cpu_temperature", "cpu_busy_frequency"
+  # cpu_metrics = []
+`
+}
+
+// Description returns a one-sentence description on the plugin.
+func (p *PowerStat) Description() string {
+	return `Intel PowerStat plugin enables monitoring of platform metrics (power, TDP) and Core metrics like temperature, power and utilization.`
+}
+
+// Gather takes in an accumulator and adds the metrics that the Input gathers.
+func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
+	p.addGlobalMetrics(acc)
+
+	if p.perCoreMetrics && len(p.msr.GetCpuIDs()) > 0 {
+		p.addPerCoreMetrics(acc)
+	}
+
+	// Gathering the first iteration of metrics was skipped for most of them because they are based on delta calculations.
+	p.skipFirstIteration = false
+
+	return nil
+}
+
+func (p *PowerStat) addPerCoreMetrics(acc telegraf.Accumulator) {
+	var wg sync.WaitGroup
+	cpuIDs := p.msr.GetCpuIDs()
+	wg.Add(len(cpuIDs))
+
+	for _, cpuID := range cpuIDs {
+		go p.addMetricsForSingleCore(cpuID, acc, &wg)
+	}
+
+	wg.Wait()
```

Comment:
How about moving this code to `Gather()`? There seems to be no further user...

Reply:
It can be done of course, but it was extracted from `Gather()` by design - to have it as small as possible.

Reply:
It's a matter of taste. It was just a suggestion. Proceed as you like.

Reply:
Thanks.